### ISSUE TITLE
runc exec command parsing

### ISSUE THREAD

**Issue by papey** [opencontainers/runc#752]
Using runc exec fails when "container command" need an option.

For example : 

```
cmd="ls -l"
runc exec ${containerid} ${cmd}
```

gives

```
Incorrect Usage.

NAME:
   runc exec - execute new process inside the container

USAGE:
   runc exec [command options] <container-id> <container command>

Where "<container-id>" is the name for the instance of the container and
"<container command>" is the command to be executed in the container.

For example, if the container is configured to run the linux ps command the
following will output a list of processes running in the container:

       # runc exec <container-id> ps

OPTIONS:
   --console                    specify the pty slave path for use with the container
   --cwd                    current working directory in the container
   --env, -e [--env option --env option]    set environment variables
   --tty, -t                    allocate a pseudo-TTY
   --user, -u                   UID (format: <uid>[:<gid>])
   --process, -p                path to the process.json
   --detach, -d                 detach from the container's process
   --pid-file                   specify the file to write the process id to
   --process-label              set the asm process label for the process commonly used with selinux
   --apparmor                   set the apparmor profile for the process
   --no-new-privs               set the no new privileges value for the process
   --cap, -c [--cap option --cap option]    add a capability to the bounding set for the process
   --no-subreaper               disable the use of the subreaper used to reap reparented processes

flag provided but not defined: -l
```

**Comment by cyphar** [comment]
It's because `runc` is treating the `-l` as a flag. You can fix this by either doing the following:

```
runc exec ${containerid} "${cmd}"
```

or

```
runc exec ${containerid} -- ${cmd}
```

**Comment by papey** [comment]
thx but runc now search for a "--" command,

```
runc exec ${containerid} -- ${cmd}
exec failed: exec: "--": executable file not found in $PATH
```

**Comment by codido** [comment]
@papey Could you give https://github.com/opencontainers/runc/pull/754 a shot?

**Comment by papey** [comment]
Yes, currently working on a PR.

**Comment by codido** [comment]
@papey Seems '--' does work, but only if it is added before the container ID:
`runc exec -- ${containerid} ${cmd}`

**Comment by jfrazelle** [comment]
```
$ sudo runc exec -t chrome ls -la
total 80
drwxr-xr-x  23 root   root    4096 Apr 14 07:40 .
drwxr-xr-x  23 root   root    4096 Apr 14 07:40 ..
-rwxr-xr-x   1 root   root       0 Apr 12 11:30 .dockerenv
drwxr-xr-x   2 root   root    4096 Apr 12 11:29 bin
drwxr-xr-x   2 root   root    4096 Mar  4 09:53 boot
drwxr-xr-x  11 root   root    4096 Apr 14 07:40 data
drwxr-xr-x   7 root   root     400 Apr 14 07:40 dev
drwxr-xr-x  48 root   root    4096 Apr 12 11:30 etc
drwxr-xr-x   2 root   root    4096 Mar  4 09:53 home
drwxr-xr-x   9 root   root    4096 Apr 12 11:29 lib
drwxr-xr-x   2 root   root    4096 Apr 12 11:29 lib64
drwxr-xr-x   2 root   root    4096 Mar 15 13:36 media
drwxr-xr-x   2 root   root    4096 Mar 15 13:36 mnt
drwxr-xr-x   3 root   root    4096 Apr 12 11:29 opt
dr-xr-xr-x 194 nobody nogroup    0 Apr 14 07:40 proc
drwx------   9 root   root    4096 Apr 14 07:40 root
drwxr-xr-x   4 root   root    4096 Apr 12 11:29 run
drwxr-xr-x   2 root   root    4096 Apr 12 11:29 sbin
drwxr-xr-x   2 root   root    4096 Apr  5 10:14 src
drwxr-xr-x   2 root   root    4096 Mar 15 13:36 srv
dr-xr-xr-x  13 nobody nogroup    0 Apr 14 07:40 sys
drwxrwxrwt   7 root   root    4096 Apr 14 07:40 tmp
drwxr-xr-x  10 root   root    4096 Apr 12 11:29 usr
drwxr-xr-x  11 root   root    4096 Apr 12 11:30 var
```

On Thu, Apr 14, 2016 at 7:24 AM, Ido Yariv <notifications@github.com> wrote:

> @papey <https://github.com/papey> Seems '--' does work, but only if it is
> added before the container ID:
> runc exec -- ${containerid} ${cmd}
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/runc/issues/752#issuecomment-209968383>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by jfrazelle** [comment]
the behavior is super odd now as to when you need and don't need `--`
looking into it

On Thu, Apr 14, 2016 at 7:41 AM, Jessica Frazelle <me@jessfraz.com> wrote:

> ```
> $ sudo runc exec -t chrome ls -la
> total 80
> drwxr-xr-x  23 root   root    4096 Apr 14 07:40 .
> drwxr-xr-x  23 root   root    4096 Apr 14 07:40 ..
> -rwxr-xr-x   1 root   root       0 Apr 12 11:30 .dockerenv
> drwxr-xr-x   2 root   root    4096 Apr 12 11:29 bin
> drwxr-xr-x   2 root   root    4096 Mar  4 09:53 boot
> drwxr-xr-x  11 root   root    4096 Apr 14 07:40 data
> drwxr-xr-x   7 root   root     400 Apr 14 07:40 dev
> drwxr-xr-x  48 root   root    4096 Apr 12 11:30 etc
> drwxr-xr-x   2 root   root    4096 Mar  4 09:53 home
> drwxr-xr-x   9 root   root    4096 Apr 12 11:29 lib
> drwxr-xr-x   2 root   root    4096 Apr 12 11:29 lib64
> drwxr-xr-x   2 root   root    4096 Mar 15 13:36 media
> drwxr-xr-x   2 root   root    4096 Mar 15 13:36 mnt
> drwxr-xr-x   3 root   root    4096 Apr 12 11:29 opt
> dr-xr-xr-x 194 nobody nogroup    0 Apr 14 07:40 proc
> drwx------   9 root   root    4096 Apr 14 07:40 root
> drwxr-xr-x   4 root   root    4096 Apr 12 11:29 run
> drwxr-xr-x   2 root   root    4096 Apr 12 11:29 sbin
> drwxr-xr-x   2 root   root    4096 Apr  5 10:14 src
> drwxr-xr-x   2 root   root    4096 Mar 15 13:36 srv
> dr-xr-xr-x  13 nobody nogroup    0 Apr 14 07:40 sys
> drwxrwxrwt   7 root   root    4096 Apr 14 07:40 tmp
> drwxr-xr-x  10 root   root    4096 Apr 12 11:29 usr
> drwxr-xr-x  11 root   root    4096 Apr 12 11:30 var
> ```
>
> On Thu, Apr 14, 2016 at 7:24 AM, Ido Yariv <notifications@github.com>
> wrote:
>
>> @papey <https://github.com/papey> Seems '--' does work, but only if it
>> is added before the container ID:
>> runc exec -- ${containerid} ${cmd}
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/opencontainers/runc/issues/752#issuecomment-209968383>
>>
>
>
>
> --
>
>
> Jessie Frazelle
> 4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
> pgp.mit.edu
> <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by jfrazelle** [comment]
seems to be more a bug in the cli lib

**Comment by jfrazelle** [comment]
Ok I see the problem, the container id is the first argument, which is not a flag, the library (correctly imo) expects the `--` to come before any arbitrary arguments, the id is kinda an arbitrary argument
unsure what to do here

**Comment by mikebrow** [comment]
While they are figuring it out... you can just use -p to point to your process.json where you can specify your arguments.  Hint:  using a quick script you can generate a new process.json with arguments and call runc with -p pointing to the new file..

**Comment by cyphar** [comment]
I think we should make a PR with the CLI library so it acts more like Python's `argparse` (that you can have mandatory positional arguments). If we have that, then we won't use `.Args().First()` which is what's causing this issue.

**Comment by keloyang** [comment]
https://github.com/urfave/cli/pull/515, Treat all flags which followd behind the last arguments as normal arguments, this will resolve the problem which runc(ps and exec) encounter.

**Comment by keloyang** [comment]
https://github.com/urfave/cli/pull/515,Treat all flags which followd behind the last arguments as normal arguments, this will resolve the problem which runc exec encounter.

**Comment by wking** [comment]
On Tue, Aug 30, 2016 at 12:08:26AM -0700, keloyang wrote:
> https://github.com/urfave/cli/pull/515…

Landed upstream via urfave/cli@3df4b7184 (urfave/cli#528) and runC
bumped past that with 1db7315 (#1051).  I haven't tested it, but this
issue can probably be closed.
